Gems/SceneProcessing/Code/Source/Generation/Components/UVsGenerator/UVsGenerateComponent.cpp

```diff
@@ -0,0 +1,254 @@
+            if (!GenerateUVsForMesh(context.GetScene(), nodeIndex, mesh, generationMethod, replaceExisting))
+            {
+                return AZ::SceneAPI::Events::ProcessingResult::Failure;
+            }
+        }
+
+        return AZ::SceneAPI::Events::ProcessingResult::Success;
+    }
+
+    bool UVsGenerateComponent::GenerateUVsForMesh(
+        AZ::SceneAPI::Containers::Scene& scene,
+        const AZ::SceneAPI::Containers::SceneGraph::NodeIndex& nodeIndex,
+        AZ::SceneAPI::DataTypes::IMeshData* meshData,
+        const AZ::SceneAPI::DataTypes::UVsGenerationMethod generationMethod,
+        const bool replaceExisting)
+    {
+        if (generationMethod == SceneAPI::DataTypes::UVsGenerationMethod::LeaveSceneDataAsIs)
+        {
+            return true;
+        }
+
+        AZ::SceneAPI::Containers::SceneGraph& graph = scene.GetGraph();
+        size_t uvSetCount = CalcUvSetCount(graph, nodeIndex);
+
+        // there might already be existing data there - see if there is.
+        AZ::SceneData::GraphData::MeshVertexUVData* dataToFill = nullptr;
+        if (uvSetCount > 0)
+        {
+            // This modifier always works on UV Set #0.
+            dataToFill = static_cast<AZ::SceneData::GraphData::MeshVertexUVData*>(FindUvData(graph, nodeIndex, 0));
+        }
+        AZStd::string currentNodeName = graph.GetNodeName(nodeIndex).GetPath();
+        if ((dataToFill) && (!replaceExisting))
+        {
+            // if there's already data, and we are not set to replace existing, do not generate data
+            AZ_Info(
+                AZ::SceneAPI::Utilities::LogWindow,
+                "Asked to generate UVs for mesh " AZ_STRING_FORMAT " but it already has UVs and 'replace existing' is not set.  Not replacing existing data.\n",
+                AZ_STRING_ARG(currentNodeName));
+            return true; // this is not an error!
+        }
+
+        if (!dataToFill)
+        {
+            if (!CreateUVsLayer(scene.GetManifest(), nodeIndex, graph, &dataToFill))
+            {
+                // the above function will emit an error if it fails.
+                return false;
+            }
+        }
+
+        bool allSuccess = true;
+        AZ_Info(
+            AZ::SceneAPI::Utilities::LogWindow,
+            "Generating UVs for " AZ_STRING_FORMAT ".\n",
+            AZ_STRING_ARG(currentNodeName));
+
+        switch (generationMethod)
+        {
+        case AZ::SceneAPI::DataTypes::UVsGenerationMethod::SphericalProjection:
+            {
+                allSuccess &= AZ::UVsGeneration::Mesh::SphericalMapping::GenerateUVsSphericalMapping(meshData, dataToFill);
+                break;
+            }
+            // for future expansion - add new methods here if you want to support additional methods of UV auto generation.
+        default:
+            {
+                AZ_Assert(false, "Unknown UVs generation method selected (%d) cannot generate UVs.\n", static_cast<AZ::u32>(generationMethod));
```

Comment:
You may get a warning here when building with clang due to %d being used instead of %u.
```suggestion
                AZ_Assert(false, "Unknown UVs generation method selected (%u) cannot generate UVs.\n", static_cast<AZ::u32>(generationMethod));
```